### DISCUSSION TITLE
Web components port of Dashboard and it's components

### DISCUSSION THREAD

**Discussion by rbjarnason** [plausible/analytics#2151] [Anything Else]
We're the [Citizens Foundation](https://citizens.is), a nonprofit civic tech organization creating open-source citizen engagement platforms since 2008, now used in 45 countries. We love Plausible and want to thank you for all the great code! :)

We use Plausible on your paid cloud service for our main nonprofit website. But we have also nearly completed the integration of Plausible into our open-source Your Priorities platform; as a part of a new Marketing and Analytics feature.

Your Priorities is built using Web Components, and we did not see a practical way to integrate the React components into our platform. So we are close to completing a production-ready port of Dashboard written in Typescript using the Lit library for Web Components, templates and reactive properties. The ported code is here: [https://github.com/CitizensFoundation/your-priorities-app/tree/master/apps/promotion_app/src/pl-components](https://github.com/CitizensFoundation/your-priorities-app/tree/master/apps/promotion_app/src/pl-components) 

There are a few new features:
- Localization support
- Support for a proxy stats URL 
- A new goal-oriented dashboard component using the stats API

Here you can see how we then extend the Plausible Dashboard web components for use in our app: [(https://github.com/CitizensFoundation/your-priorities-app/tree/master/apps/promotion_app/src/yp-analytics)](https://github.com/CitizensFoundation/your-priorities-app/tree/master/apps/promotion_app/src/yp-analytics). 

We just wanted to let you know about this work; we don't expect you to ditch React. The Web Components standards approach works with any framework (or no framework, as is our case). We have been using Web Components since 2014 and are very happy developers. Let me know if you are interested in a PR! ;) but otherwise, I think this new code will live within our codebase for the time being.

Here are dev screenshots:
![image](https://user-images.githubusercontent.com/43699/187480819-1c91031b-19e1-4ca5-8f1b-21af477edc09.png)

![image](https://user-images.githubusercontent.com/43699/187485391-464976c3-1066-4e6f-a4cd-8b667fac63ab.png)

**Comment by rbjarnason** [comment]
This week, we launched the Plausible integrated solution in production on our open-source citizen engagement platform, Your Priorities, for example, here: https://yrpri.org/

Here is the Web Components port of the Plausible Dashboard components:
https://github.com/CitizensFoundation/your-priorities-app/tree/master/apps/promotion_app/src/pl-components

Here are some screenshots from a participatory budgeting project in Reykjavik that launched today, with Icelandic localization.

![image](https://user-images.githubusercontent.com/43699/191826580-8b8104cd-cf15-4367-ba94-f2b2232fe771.png)

![image](https://user-images.githubusercontent.com/43699/191826659-682df082-b2ba-4d9a-890f-95e684719e62.png)

![image](https://user-images.githubusercontent.com/43699/191826736-78239520-b40e-4c18-a353-4185ba88034b.png)

**Comment by ukutaht** [comment]
Wow, this is incredible work by your team!

I should look into Web Components and what benefit they bring :)

**Reply by rbjarnason** [reply]
Thanks! :)

Of course, I'm pretty biased towards the Web Components standard, having used it since 2014! ;) But here is how I think it could help you, especially your enterprise customers who, while using your API backend as a service, might want to integrate the Plausible UI into their websites written in all possible languages and frameworks:

* Shadow DOM. The web component versions of the Plausible components can be dropped into any website or layout; nothing from the outside can mess with the Plausible CSS, and it can't affect the website's CSS either.
* Standards-based. You could have an npm install-based system for the Plausible Dashboard components that can be dropped into any Javascript-based app, including React apps.
* You can offer a framework-independent UI SDK for your enterprise customers. We are already using Plausible as an SDK. Our components are sub-classes of the Plausible components; we compose HTML by combining different Plausible dashboard components and adding our components. Check out how we use this here: https://github.com/CitizensFoundation/your-priorities-app/tree/master/apps/promotion_app/src/yp-analytics

**Comment by rbjarnason** [comment]
FYI - Here is a new blog post about the Plausible integration with Your Priorities: https://citizens.is/2022/10/16/new-analytics-and-promotion-tool/
